### PR TITLE
Remove some copies from the queue binding implementation

### PR DESCRIPTION
**Pull request by a-robinson**
**Remove some copies from the queue binding's send() method**

By not copying data out of perfectly good Strings into new heap arrays
when we don't need to.

**Remove a copy when calling queue.sendBatch with "bytes" content**

Because we synchronously consume the arraybuffer in order to base64-
encode it further down in sendBatch, there's no point in making a full
copy of it.

In theory we could just always call `BufferSource::detach()` to avoid
ever needing a copy, but I'm a little worried about the unintuitive
behavior of an arraybuffer being emptied out whenever you pass it into a
queue method. While it'd be really nice for performance, I'd hate to see
it cause a serious customer bug.

### CODE REVIEW

File: src/workerd/api/queue.c++ [hunk 1]
```diff
@@ -63,10 +82,24 @@ kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::Stri
...
     jsg::BufferSource source(js, body);
-    return kj::heapArray(source.asArrayPtr());
+    if (forSynchronousUse) {
+      // If we know the data will be consumed synchronously, we can avoid copying it.
+      Serialized result;
+      result.data = source.asArrayPtr();
+      result.own = kj::mv(source);
+      return kj::mv(result);
+    }
+    kj::Array<kj::byte> bytes = kj::heapArray(source.asArrayPtr());
```
Comment (a-robinson) [inline src/workerd/api/queue.c++:92]: cc @jasnell in case you want to assuage my fears that doing a `source.detach()` would be ok here rather than always copying the input data, but I notice that we didn't bring https://github.com/cloudflare/workerd/pull/313 to a clear conclusion so I'm not sure what our current opinion is on whether APIs should consume ArrayBuffers that are passed to them.

File: src/workerd/api/queue.c++ [hunk 2]
```diff
@@ -24,18 +24,34 @@ kj::StringPtr validateContentType(kj::StringPtr contentType) {
...
+  result.own = kj::mv(bytes);
+  return kj::mv(result);
 }
 
-kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::StringPtr contentType) {
+Serialized serialize(
+    jsg::Lock& js,
+    v8::Local<v8::Value> body,
+    kj::StringPtr contentType,
+    bool forSynchronousUse = false) {
```
Comment (jasnell) [inline src/workerd/api/queue.c++:54]: nit: generally not a fan of bool args like this as it makes the call site a bit less obvious. Would prefer an enum here instead but not blocking.

File: src/workerd/api/queue.c++ (hunk 1 above)
Comment (jasnell) [reply src/workerd/api/queue.c++:92]: For existing APIs, I think we're stuck with the copy. Calling detach has an observable effect on the call site side that could introduce bugs in the user code.

File: src/workerd/api/queue.c++ (hunk 2 above)
Comment (a-robinson) [reply src/workerd/api/queue.c++]: Yeah, that's pretty much always clearer, and it's fair to call me out on it. Updated.

File: src/workerd/api/queue.c++ (hunk 1 above)
Comment (a-robinson) [reply src/workerd/api/queue.c++:99]: This is still a new API, though -- this code path is only used with non-default `contentType`s, and `contentType` support has not yet been released AFAIK (correct me if that's wrong, @jbwcloudflare ). Does that change how we feel about it?

Comment (jbwcloudflare) [reply src/workerd/api/queue.c++:99]: Depends what you mean. The code is released and not behind a compatibility flag, so technically anyone could be using `contentType` right now. But we have not many any public announcement or documentation of the feature just yet.

IMO it's low risk and we should be fine to change it now.

Comment (a-robinson) [reply src/workerd/api/queue.c++:99]: Yeah I don't have any big concerns about breaking someone who has managed to already start using the feature and start depending on being able to continue using ArrayBuffers that they pass in with type "bytes".

But given that we're ok making the change, do we actually want to? That's my main question. It feels weird and possibly error-prone to me to consume the array, but if we're gonna start doing it in more APIs over time then we might as well conform.

Comment (a-robinson) [reply src/workerd/api/queue.c++]: Done in https://github.com/cloudflare/workerd/pull/772/commits/64ff5fcdce5317ab8283bace62d8a48ffeba1c5c

**Comment by a-robinson** [comment]
The latest force-push was just to rebase on main after two weeks of changes had built up. The only actual code change was the new commit to use `detach()`.

There's no other changes needed here as far as I know, so if someone approves then I'll plan on merging it once it's green.

**Review by jasnell — APPROVED** [review]
